File: RuyiSDKCpp/RuyiNet/Task/RuyiNetTaskQueue.cpp
#include "RuyiNetTaskQueue.h"

namespace Ruyi
{
	RuyiNetTaskQueue::~RuyiNetTaskQueue()
	{
		while (mTaskQueue.size() > 0)
		{
			delete mTaskQueue.front();
			mTaskQueue.pop();
		}

		if (mCurrentTask != nullptr)
		{
			delete mCurrentTask;
			mCurrentTask = nullptr;
		}
	}

	void RuyiNetTaskQueue::Enqueue(RuyiNetTaskBase * task)
	{
		mTaskQueue.push(task);
	}

	void RuyiNetTaskQueue::Update()
	{
		if (mCurrentTask != nullptr)
		{
			if (mCurrentTask->GetCompleted())
			{
				mCurrentTask->OnResponse();
				mTaskThread = nullptr;

				delete mCurrentTask;
				mCurrentTask = nullptr;
			}
		}
		else if (mTaskQueue.size() > 0)
		{
			mCurrentTask = std::move(mTaskQueue.front());
			mTaskQueue.pop();
			mTaskThread = std::make_unique<std::thread>(std::bind(&RuyiNetTaskBase::Execute, mCurrentTask));
		}
	}
}

File: RuyiSDK/RuyiNet/Task/RuyiNetTaskQueue.cs
using System.Collections.Generic;
using System.Threading;

namespace Ruyi.SDK.Online
{
    internal class RuyiNetTaskQueue
    {
        public RuyiNetTaskQueue()
        {
            mTaskQueue = new Queue<RuyiNetTaskBase>();
        }

        public void Enqueue(RuyiNetTaskBase task)
        {
            mTaskQueue.Enqueue(task);
        }

        public void Update()
        {
            if (mCurrentTask != null)
            {
                if (mCurrentTask.Completed)
                {
                    mCurrentTask.OnResponse();

                    mTaskThread = null;
                    mCurrentTask = null;
                }
            }
            else if (mTaskQueue.Count > 0)
            {
                mCurrentTask = mTaskQueue.Dequeue();
                mTaskThread = new Thread(new ThreadStart(() => { mCurrentTask.Execute(); }));
                mTaskThread.Start();
            }
        }

        public int Work
        {
            get
            {
                if (mCurrentTask != null)
                {
                    return mTaskQueue.Count + 1;
                }

                return mTaskQueue.Count;
            }
        }

        private Thread mTaskThread;
        private Queue<RuyiNetTaskBase> mTaskQueue;
        private RuyiNetTaskBase mCurrentTask;
    }
}
